tests/DirEnt.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Main (main) where
+
+import Control.Exception                (bracket, finally)
+import Foreign.C.String                 (peekCString)
+import System.Exit
+import System.Posix.Directory
+import System.Posix.Directory.Internals
+import System.Process                   (system)
+
+system_x :: String -> IO ExitCode
+system_x cmd = system $ "set -x; " ++ cmd
+
+onFailure :: IO ExitCode -> (ExitCode -> IO ()) -> IO ()
+action `onFailure` after = action >>= \case
+    ExitSuccess -> return ()
+    ec          -> after ec
+infixr 9 `onFailure`
+
+prepareTest :: IO ()
+prepareTest = do
+    system_x "cc --version" `onFailure` exitWith
```

Comment:
I think this is highly non-portable. I'd prefer to remove the c-tests altogether, do "static" checks in `configure.ac` and do additional runtime checks (whether the kernel just returns `DT_UNKNOWN`) in the Haskell code itself.

If we really want to run C code, we can compile it properly through cabal and use the C FFI to run the appropriate function. I do not understand why we compile manually in the test suite and produce a binary here.

Reply:
> When d_type is defined and the filesystem supports non-zero values for it,

Yes, these absolutely can be tested for in `configure.ac`. I think one of my comments got swallowed by GitHub.

* https://autotools.info/autoconf/custom-tests.html

Reply:
> Assume that for Linux, GitHub CI and Hackage CI always make `.` have `d_type` `DT_DIR`, and test exactly that with `dirEntType`, and only run it under Linux. If it returns `UnknownType` or some other value, fail the test noisily, running `mount(8)`, and deal with that when that happens.

I think it's good enough and strikes a good balance. Although I do not get what `mount(8)` reference is about.

Reply:
It's technically a non-portable test and so I think it should be guarded behind a cabal flag. Maybe `non-portable-tests`?

Reply:
@hasufell I disabled it for non-Linux in `unix.cabal` but maybe that's not what you're talking about.  I set "Allow edits by maintainers" so please feel free to add any Cabal flags.